components/layout_2020/table/construct.rs

```diff
@@ -4,16 +4,133 @@ use crate::dom_traversal::{
     BoxSlot, Contents, NodeAndStyleInfo, NodeExt, NonReplacedContents, TraversalHandler,
 };
 use crate::style_ext::{DisplayGeneratingBox, DisplayInternal};
+use script_layout_interface::wrapper_traits::ThreadSafeLayoutNode;
 use std::borrow::Cow;
+use std::cmp;
 use std::convert::TryFrom;
 use style::values::specified::text::TextDecorationLine;
 
 #[derive(Debug, Serialize, Default)]
+/// A map of table slots to cells
 pub(crate) struct TableSlots {
     rows: Vec<TableSlotsRow>,
 }
 
-#[derive(Debug, Serialize)]
+impl TableSlots {
+    /// Get the slot at (x, y)
+    pub fn get(&self, x: usize, y: usize) -> Option<&TableSlot> {
+        self.rows.get(y)?.cells.get(x)
+    }
+
+    fn insert(&mut self, slot: TableSlot) {
+        let y = self.rows.len() - 1;
+        self.rows[y].cells.push(slot)
+    }
+
+    /// Convenience method for get() that returns a SlotAndLocation
+    fn get_loc(&self, x: usize, y: usize) -> Option<SlotAndLocation> {
+        self.rows
+            .get(y)?
+            .cells
+            .get(x)
+            .map(|slot| SlotAndLocation { slot, x, y })
+    }
+
+    /// Get the slot from the previous row at x
+    fn get_above(&self, x: usize) -> Option<SlotAndLocation> {
+        if self.rows.len() > 1 {
+            self.get_loc(x, self.rows.len() - 2)
+        } else {
+            None
+        }
+    }
+
+    /// Given a slot and location, find the originating TableSlot::Cell.
+    /// In the case of a table model error, there may be multiple, returning
+    /// the oldest cell first
+    fn resolve_slot<'a>(
+        &'a self,
+        location: SlotAndLocation<'a>,
+    ) -> (SlotAndLocation<'a>, Vec<SlotAndLocation<'a>>) {
+        match *location.slot {
+            TableSlot::Cell { .. } => (location, Vec::new()),
+            TableSlot::Spanned(x, y) => (
+                self.get_loc(location.x - x, location.y - y)
+                    .expect("Spanned slot reference must resolve"),
+                Vec::new(),
+            ),
+            TableSlot::MultiSpanned(ref vec) => {
+                let mut v: Vec<_> = vec
+                    .iter()
+                    .map(|(x, y)| {
+                        self.get_loc(location.x - x, location.y - y)
+                            .expect("Spanned slot reference must resolve")
+                    })
+                    .collect();
+                (v.pop().unwrap(), v)
+            },
+        }
+    }
+
+    /// If (x, y) is spanned by an already resolved `spanner`, return offsets for a TableSlot::Spanned() for it
+    fn spanned_slot_single(
+        &self,
+        x: usize,
+        y: usize,
+        spanner: SlotAndLocation,
+    ) -> Option<(usize, usize)> {
+        if let TableSlot::Cell { width, height, .. } = spanner.slot {
+            if x >= spanner.x && y >= spanner.y && x < spanner.x + *width {
+                if *height == 0 || y < spanner.y + *height {
```

Comment:
Why two nested `if`s here?

Reply:
One is an && chain, the other is an || chain, mixing them looks confusing